Remove legacy customization system

unlocomqx opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Removes code related to the legacy customization system
| Type?         | refacto
| Category?     | CO
| BC breaks?    | yes
| Deprecations? | yes
| Fixed ticket? | Fixes https://github.com/PrestaShop/PrestaShop/issues/10912
| How to test?  | This PR removes the significance of the customization quantity because each customization has its row in the cart so it has the quantity of the respective row. Please test this with this PR https://github.com/PrestaShop/classic-theme/pull/25 (classic theme).<br/>It introduces a fix to display the customization quantity normally in the order details in the customer account
| Related PRs       | https://github.com/PrestaShop/classic-theme/pull/25
| Possible impacts? | N/A.

This PR breaks customizations saved using the old customization system (-> saved using PrestaShop 1.6)
It also changes the order detail data as it adds the customization price & qty directly in the order detail row, this would break the display of orders already saved with a customization price & qty different than 0

## Deprecations

### \ProductCore::addCustomizationPrice
This method was used to assign pricing information to multiple customizations of multiple products. Since each cart row can now have a customization, the pricing of the customization is the same as the cart row pricing. Hence this method is no longer needed. The pricing is handled by `\ProductCore::getPriceStatic`.


### \ProductCore::addProductCustomizationPrice
Similar to `\ProductCore::addCustomizationPrice` but used for a single product.


### \CustomizationCore::countQuantityByCart
This method was needed in order to get the quantity of each customization in cart and then use that information to validate the order cancellation to make sure we don't cancel a bigger quantity than the customization quantity that is in cart.
This is no longer needed because we only care about canceling cart lines as customization no longer have a dedicated quantity.

### \CustomizationCore::$quantity
We currently only use the quantity from the corresponding`cart_product` row.


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/12422)
<!-- Reviewable:end -->

unlocomqx committed:
BO: simplify admin carts view

unlocomqx committed:
BO: simplify admin orders

unlocomqx committed:
BO: remove addProductCustomizationPrice call from admin carts

unlocomqx committed:
BO: remove class to addProductCustomizationPrice

unlocomqx committed:
BO: simplify the invoice

unlocomqx committed:
BO: add customization price & weight to order detail

unlocomqx committed:
add deprecated tags to addProductCustomizationPrice & addCustomizationPrice

unlocomqx committed:
FO: remove customization quantity management in cart

unlocomqx committed:
fix code style

matks commented:
This seems to make sense however we need to ask people that have been working on this project longer than me whether this is not going to break some behavior.
@Quetzacoalt91 @mickaelandrieu ? 😄

unlocomqx commented:
> This seems to make sense however we need to ask people that have been working on this project longer than me whether this is not going to break some behavior.
> @Quetzacoalt91 @mickaelandrieu ?

It's quite a scary jump tbh but I think it's worth it because the legacy system is becoming a source of bugs as in #10912

PierreRambaud commented:
ping @Quetzacoalt91 @mickaelandrieu  still waiting for your knowledge guys :D

matks commented:
Hi everyone [following our review guidelines](https://devdocs.prestashop.com/1.7/project/maintainers-guide/reviewing-pull-requests/) "Pull Requests may be closed after 30 days of inactivity"

Since this PR did not get any activity since December 2019, I close it.

Please note this does not mean it's a final "won't merge" 😉  ; **if you come back we can reopen it and update it**

Thank you for the contribution

unlocomqx commented:
@matks Can you reopen this?
Imo, the old customization system shouldn't go into 8.0
After many years of after-sales, I didn't see any customer who needed the old system at all.
Supporting both doesn't make sense
It will always make bug fixing and adding new features more difficult.

I can rebase and rework this PR if needed

matks commented:
Reopened 😄

matks commented:
So I think a rebase will be needed after 2 years :/ maybe a complete rewrite is easier? so many code has changed since then, maybe rebase will be too long

unlocomqx commented:
Yes a rewrite is much easier at this point 👍

unlocomqx committed:
Remove Cart::_updateCustomizationQuantity

This is a protected method so it can be removed directly

unlocomqx committed:
Ignore the quantity in customization table

Only the quantity from cart_product will be used, because each customization has a separate cart row

unlocomqx committed:
Assume we are using the new customization method

This check is no longer necessary

unlocomqx committed:
Set customization in_cart after adding it to cart

unlocomqx commented:
I rebased it (made relatively easy by phpstorm)
I'll re-test everything and check if there are new areas that need to be adapted

unlocomqx committed:
Remove code that is specific to customization quantity

The customization quantity will always be the same as the cart_quantity value

unlocomqx committed:
Reformat code

atomiix commented:
Hi @unlocomqx, it looks like tests are failing, can you solve them?

unlocomqx commented:
@atomiix Yes, I will work on it
Cheers

unlocomqx committed:
Update customization attribute and delivery address

unlocomqx committed:
Simplify order canceling

The customization doesn't need to be canceled because each customization has its own order_detail row

unlocomqx committed:
Reformat code

atomiix commented:
@unlocomqx good work! However, like @matks said, I'm afraid it breaks an expected behavior, I'll try to ping @Quetzacoalt91 one more time 😅
Also, maybe I'm wrong but it looks like we could get rid of some fields in the `customization` table, like `quantity*` fields?
We'll have to add an SQL migration script too I guess?

unlocomqx commented:
As far as I tested, I didn't find any breakage

I didn't want to touch the DB before discussing it with you.
We can drop the 
quantity
quantity_refunded
quantity_returned

There are other fields that can also be dropped but it requires a bit wider refactoring
Some state is currently tracked both in cart_product and customization, like `id_address_delivery` and `id_product_attribute` etc

MeKeyCool commented on classes/Cart.php:
```diff
@@ -1653,6 +1654,14 @@ public function updateQty(
+                    'id_address_delivery' => (int) $id_address_delivery,
+                    'in_cart' => 1,
+                ], '`id_customization` = ' . (int) $id_customization);
+            }
```
`int` type juggling then `string` one ...
At l.1517 somebody started to ensure all types to remove all those type jugglings. It would be nice to clean a bit more.

(non blocking comment)

unlocomqx replied:
Thanks for your comment
I didn't detect where the issue is exactly, maybe it's obvious but I can't see it

MeKeyCool replied:
It is not really an issue. Reading this part i just saw type juggling everywhere and even a line like : 
```php
$string_variable . (int) $another_variable
``` 
(double type juggling xD)

And looking a bit more the function, i saw that : 
```php
$quantity = (int) $quantity;
$id_product = (int) $id_product;
$id_product_attribute = (int) $id_product_attribute;
```
Somebody already started to clean a bit.

But i think it is not linked to your work so you can forget my comment if you don't have time for this.

unlocomqx replied:
Can you link to code that has already been cleaned, i'm just curious :))

MeKeyCool replied:
In the same function you can see that there are two different ways to manage type juggling :

l. 1517
```php
        $quantity = (int) $quantity;
        $id_product = (int) $id_product;
        $id_product_attribute = (int) $id_product_attribute;
```

l. 1575
```php
        /* Check if the product is already in the cart */
        $cartProductQuantity = $this->getProductQuantity(
            $id_product,
            $id_product_attribute,
            (int) $id_customization,
            (int) $id_address_delivery
        );
```

unlocomqx replied:
Now I see what you mean, thank you for your remarks
I fixed the are that I touched 
https://github.com/PrestaShop/PrestaShop/pull/12422/commits/4f5c0800f7277d2fffa59e62c195259c368b51b1
Tbh, I don't want to touch other areas because it will make the PR scope even wider

MeKeyCool commented on classes/Product.php:
```diff
@@ -6061,6 +6061,8 @@ public static function getAllCustomizedDatas($id_cart, $id_lang = null, $only_in
     }
 
     /**
+     * @deprecated 8.0.0
```
More detailed comments would be nice i guess :thinking: (can't be sure cause i don't know code base enough ^^')

https://devdocs.prestashop.com/1.7/development/coding-standards/#deprecations

unlocomqx replied:
Improved in https://github.com/PrestaShop/PrestaShop/pull/12422/commits/f1ee0055bb886d688dd124f1f594ad69973697cd

MeKeyCool reviewed (commented):
This part of the code is too complicate for me to had useful comments ^^'

Thanks for the time you take to clean this ! ^^'

unlocomqx committed:
Improve deprecation comments

atomiix commented on pdf/invoice.product-tab.tpl:
```diff
@@ -138,15 +138,11 @@
+					<td class="center"></td>
 
           {assign var=end value=($layout._colCount-3)}
           {for $var=0 to $end}
-            <td class="center">
-              --
-            </td>
+						<td class="center"></td>
```
Just to be a bit picky, can you fix the indentation here, please?

unlocomqx replied:
Tabs have slipped in there somehow, fixed in https://github.com/PrestaShop/PrestaShop/pull/12422/commits/b8e71593cf9c85e380de95b57cb6bea1d3c94927

atomiix reviewed (dismissed):
Just a small style improvement feedback. Good PR overall, I think that with a good QA campaign on the PR, we'll be safe to merge it!

unlocomqx committed:
Fix indentation

unlocomqx committed:
Clean up type juggling

kpodemski commented on classes/Cart.php:
```diff
@@ -1521,7 +1522,9 @@ public function updateQty(
-        $product = new Product($id_product, false, (int) Configuration::get('PS_LANG_DEFAULT'), $shop->id);
+        $id_customization = (int) $id_customization;
+        $id_address_delivery = (int) $id_address_delivery;
+        $product = new Product($id_product, false, Configuration::get('PS_LANG_DEFAULT'), $shop->id);
```
```suggestion
        $product = new Product($id_product, false, (int) Configuration::get('PS_LANG_DEFAULT'), $shop->id);
```

@unlocomqx PHPStan is not happy about having not integer value there :-)

kpodemski commented:
Hello @unlocomqx 

Only one little change left.

I see that your issue has been marked as a duplicate of:
https://github.com/PrestaShop/PrestaShop/issues/10912

If that's true, you can update the description to have "Fixes #10912" 👍 

Thank you for your work on this PR. Great work 💪

unlocomqx commented:
@kpodemski Yes, my issue was a duplicate of #10912 
I confirm that this PR fixes it :))

Thank you

kpodemski commented:
Hey @unlocomqx 

Could you do one more rebase, please? :( Something has been merged before your PR.

kpodemski commented on classes/Cart.php:
```diff
@@ -1521,7 +1522,9 @@ public function updateQty(
-        $product = new Product($id_product, false, (int) Configuration::get('PS_LANG_DEFAULT'), $shop->id);
+        $id_customization = (int) $id_customization;
+        $id_address_delivery = (int) $id_address_delivery;
+        $product = new Product($id_product, false, Configuration::get('PS_LANG_DEFAULT'), $shop->id);
```
```suggestion
        $product = new Product($id_product, false, (int) Configuration::get('PS_LANG_DEFAULT'), $shop->id);
```

@unlocomqx

unlocomqx replied:
Resolved in https://github.com/PrestaShop/PrestaShop/pull/12422/commits/1f63ddc263b240e5cd201a020f90b7e967abf2f2

unlocomqx committed:
Cast default language id

Progi1984 commented on classes/Customization.php:
```diff
@@ -277,6 +277,8 @@ public static function retrieveQuantitiesFromIds($idsCustomizations)
 
     /**
+     * @deprecated 8.0.0 This was needed to validate order canceling
+     *
```
Could you deprecate the method using the method explained in devdocs (https://devdocs.prestashop.com/1.7/development/coding-standards/#deprecations) ?

unlocomqx replied:
Resolved in https://github.com/PrestaShop/PrestaShop/pull/12422/commits/3d20bbc2854ff0063ff10280afdcb0f1464f5d3b

Progi1984 commented on classes/Product.php:
```diff
@@ -6091,6 +6091,8 @@ public static function getAllCustomizedDatas($id_cart, $id_lang = null, $only_in
 
     /**
+     * @deprecated since 8.0.0, the customization price impact is already included in Product::getPriceStatic
+     *
```
Could you deprecate the method using the method explained in devdocs (https://devdocs.prestashop.com/1.7/development/coding-standards/#deprecations) ?

unlocomqx replied:
Resolved in https://github.com/PrestaShop/PrestaShop/pull/12422/commits/3d20bbc2854ff0063ff10280afdcb0f1464f5d3b

Progi1984 commented on classes/Product.php:
```diff
@@ -6151,6 +6153,8 @@ public static function addCustomizationPrice(&$products, &$customized_datas)
     }
 
     /**
+     * @deprecated since 8.0.0, the customization price impact is already included in Product::getPriceStatic
```
Could you deprecate the method using the method explained in devdocs (https://devdocs.prestashop.com/1.7/development/coding-standards/#deprecations) ?

unlocomqx replied:
Resolved in https://github.com/PrestaShop/PrestaShop/pull/12422/commits/3d20bbc2854ff0063ff10280afdcb0f1464f5d3b
I'll add documentation about the deprecated methods

Progi1984 commented on src/Adapter/Order/OrderProductQuantityUpdater.php:
```diff
@@ -505,23 +503,15 @@ protected function reinjectQuantity(
      * @throws OrderException
      */
-    private function updateCustomizationOnProductDelete(Order $order, OrderDetail $orderDetail, int $oldQuantity): void
+    private function deleteProductCustomization($id_customization): void
```
```suggestion
    private function deleteProductCustomization(int $id_customization): void
```

unlocomqx replied:
Resolved in https://github.com/PrestaShop/PrestaShop/pull/12422/commits/517ebc0c1e4aeec3b6053e28d45757db27194f0b

Progi1984 reviewed (commented):
@unlocomqx Thank you for your contribution. I added some comments about deprecations (+ Could you document in your PR methods you have deprecated ?). Thanks

unlocomqx committed:
Add deprecation error triggers

unlocomqx committed:
Add parameter type

unlocomqx committed:
Cast id_customization to integer

marwachelly reviewed (commented):
Hello @unlocomqx,

I tested the PR and I detected a regression related to he PR following the senario bellow :

1. Create an order from BO
2. Check the order details from FO 
3. The product quantity is not correct

With PR:


https://user-images.githubusercontent.com/89604350/162741878-e66c5592-9e0a-4139-b26f-1ac4b78d2939.mp4

 Without PR:

https://user-images.githubusercontent.com/89604350/162741659-5ee1d102-f6b8-4b7e-beff-5778fc934023.mp4

Could you please check?

Thanks,

unlocomqx commented:
@marwachelly A fix needs to be applied on the classic theme
I made the fix here https://github.com/unlocomqx/classic-theme/commit/cd69c04f7ba7b5d5111c4523b02499e8560ed804

![screenshot-000286](https://user-images.githubusercontent.com/793712/162746763-0fe1798e-a9e8-474a-a531-381d99bd9fa3.png)


I will submit a PR for the classic theme as well
I will also check the order return template

unlocomqx committed:
Simplify the order return feature by removing customization specific logic

atomiix commented on classes/Customization.php:
```diff
@@ -287,6 +287,14 @@ public static function retrieveQuantitiesFromIds($idsCustomizations)
     {
+        @trigger_error(
+            sprintf(
+                '%s is deprecated since version 8.0.0. This method is no longer used to validate order cancellation.',
```
I'm not sure of this deprecated message. I think that this method can be called for something else that order cancellation validation 🤔

unlocomqx replied:
Yes, you are right. I will remove this deprecation.

unlocomqx replied:
Resolved in https://github.com/PrestaShop/PrestaShop/pull/12422/commits/facb0227b0dfcdbdd84d5e4aa9c373a448d24e4a

atomiix commented on classes/order/OrderReturn.php:
```diff
@@ -62,8 +62,15 @@ class OrderReturnCore extends ObjectModel
     ];
 
-    public function addReturnDetail($order_detail_list, $product_qty_list, $customization_ids, $customization_qty_input)
+    public function addReturnDetail($order_detail_list, $product_qty_list)
     {
+        if (func_num_args() > 2) {
+            @trigger_error(
```
Removing arguments is BC break, I suggest to do it like that to deprecate it:
```suggestion
    public function addReturnDetail($order_detail_list, $product_qty_list, $customization_ids = null, $customization_qty_input = null)
    {
        if ($customization_ids !== null || $customization_qty_input !== null) {
            @trigger_error(
```

unlocomqx replied:
Resolved in https://github.com/PrestaShop/PrestaShop/pull/12422/commits/ba5f7ae6de13acbf71a11cd67c5a7e0e9cd912b4

atomiix commented on classes/order/OrderReturn.php:
```diff
@@ -74,20 +81,17 @@ public function addReturnDetail($order_detail_list, $product_qty_list, $customiz
     }
 
-    public function checkEnoughProduct($order_detail_list, $product_qty_list, $customization_ids, $customization_qty_input)
+    public function checkEnoughProduct($order_detail_list, $product_qty_list)
     {
+        if (func_num_args() > 2) {
+            @trigger_error(
```
Same here:
```suggestion
    public function checkEnoughProduct($order_detail_list, $product_qty_list, $customization_ids = null, $customization_qty_input = null)
    {
        if ($customization_ids !== null || $customization_qty_input !== null) {
            @trigger_error(
```

unlocomqx replied:
Resolved in https://github.com/PrestaShop/PrestaShop/pull/12422/commits/ba5f7ae6de13acbf71a11cd67c5a7e0e9cd912b4

unlocomqx committed:
Remove deprecation of countQuantityByCart

unlocomqx commented:
@QA Team, Please test this with this PR https://github.com/PrestaShop/classic-theme/pull/25
It introduces a fix to display the customization quantity normally in the order details in the customer account

unlocomqx committed:
Improve deprecation logic

marwachelly commented:
Hello @unlocomqx,

I tested both PR and the issue of quantity is well fixed but I noticed a small difference in the invoice when ordering a customized product with different quantity and different customization:
**without PR:**
![image](https://user-images.githubusercontent.com/89604350/164264090-86d9eea8-3262-41d2-aa30-76152139c57c.png)

**with PR:**

![image](https://user-images.githubusercontent.com/89604350/164264384-0e9d086e-f1c8-4ebd-b210-1ef3e424457f.png)

Could you please check and feedback ?

Thanks ,

unlocomqx commented:
@marwachelly Thanks for checking, I basically removed the duplicated information because the customization qty and product line qty are the same.
![screenshot-000316](https://user-images.githubusercontent.com/793712/164298860-44bae1cd-fcc9-4f0b-a968-53753717146c.png)
So the cells that I removed don't add any information and are better emptied.

unlocomqx commented:
Hi @marwachelly 👋 , it's still marked as waiting for author, please check when you get to it

Hlavtox commented:
Only dangers I see are `addCustomizationPrice` and `addProductCustomizationPrice`.
@Matt75 Can these functions be used by modules to add some kind of custom extra charges?

matthieu-rolland commented:
hi @marwachelly ! I think you got an answer, approving this PR or not is up to you now ;)

marwachelly reviewed (commented):
Hello @PrestaShop/product-team,

Did you agree with the [change](https://github.com/PrestaShop/PrestaShop/pull/12422#issuecomment-1104308043) done by @unlocomqx ?

Thanks,

Gruttershop commented:
![image](https://user-images.githubusercontent.com/116812745/208751061-b380356b-7b54-4f46-8306-1a716c681f37.png)

How can we change the Fax? or edit under fax also KVK number and also Vat number. it's only 2 extra line's. i crash my site again. just show my where i must be. if you don't want to use i need.

jf-viguier commented:
IMO, This PR is not enough document. Is the customization price/weight addition is removed ?
I use a lot the price and weight columns of customized_data to modulate product price regarding the customization.

Is the "legacy customization system" is the full customization system or only few methods ?

Hlavtox commented:
@jf-viguier It doesn't remove any functionality, don't worry.

In 1.6, there was just one line per product in the cart and all the info about customization quantities was stored in `ps_customization`. In 1.7, this is changed and each customized product has it's own line in `ps_cart_product` table, along with a new column in that table `id_customization`.

Price and weight addons and other data is in `ps_customized_data`, that is staying of course. 👍

Hlavtox commented:
@MatShir Can you please ✅ the comment of @marwachelly so we can get this merged?

@unlocomqx We need to resolve that conflict brother. :(

unlocomqx commented:
@jf-viguier Check my forum post that I did back then
https://www.prestashop.com/forums/topic/549666-guide-the-new-customization-system-in-prestashop-17/

As @Hlavtox mentioned, this will keep the 1.7 customization and remove the 1.6 retro-compatibility code to make the code simpler.

@Hlavtox I will rebase it shortly

kpodemski commented:
@Hlavtox, unfortunately, we need to wait for v9 to have it merged, it has to wait

kpodemski commented on classes/Product.php:
```diff
@@ -6090,11 +6090,21 @@ public static function getAllCustomizedDatas($id_cart, $id_lang = null, $only_in
     {
+        @trigger_error(
+            sprintf(
+                '%s is deprecated since version 8.0.0. The customization price impact is already included in Product::getPriceStatic.',
```
```suggestion
                '%s is deprecated since version 9.0.0. The customization price impact is already included in Product::getPriceStatic.',
```

kpodemski commented on classes/Product.php:
```diff
@@ -6090,11 +6090,21 @@ public static function getAllCustomizedDatas($id_cart, $id_lang = null, $only_in
     }
 
     /**
+     * @deprecated since 8.0.0, the customization price impact is already included in Product::getPriceStatic.
```
```suggestion
     * @deprecated since 9.0.0, the customization price impact is already included in Product::getPriceStatic.
```

kpodemski commented on classes/Product.php:
```diff
@@ -6150,13 +6160,23 @@ public static function addCustomizationPrice(&$products, &$customized_datas)
     }
 
     /**
+     * @deprecated since 8.0.0, the customization price impact is already included in Product::getPriceStatic.
```
```suggestion
     * @deprecated since 9.0.0, the customization price impact is already included in Product::getPriceStatic.
```

kpodemski commented on classes/Product.php:
```diff
@@ -6150,13 +6160,23 @@ public static function addCustomizationPrice(&$products, &$customized_datas)
     {
+        @trigger_error(
+            sprintf(
+                '%s is deprecated since version 8.0.0. The customization price impact is already included in Product::getPriceStatic.',
```
```suggestion
                '%s is deprecated since version 9.0.0. The customization price impact is already included in Product::getPriceStatic.',
```

kpodemski commented on classes/order/OrderReturn.php:
```diff
@@ -74,20 +81,17 @@ public function addReturnDetail($order_detail_list, $product_qty_list, $customiz
     {
+        if ($customization_ids !== null || $customization_qty_input !== null) {
+            @trigger_error(
+                'Passing customization infos is deprecated since version 8.0.0. The customization is already included in the order details.',
```
```suggestion
                'Passing customization infos is deprecated since version 9.0.0. The customization is already included in the order details.',
```

kpodemski reviewed (commented):
It's time, finally, to have this PR merged :D 

A few comments as it'll target v9 instead of v8 :)

unlocomqx committed:
Change base version of deprecations

Hlavtox commented:
@unlocomqx Congratulations on advancing to the finals of the longest open PR contest

unlocomqx commented:
We can make this PR a birthday surprise 🎂 😄 
I edited the files and changed ´8.0.0´ to ´9.0.0´
Thank you

jf-viguier commented:
Happy birthday young PR ;-)

kpodemski reviewed (changes requested):
Hi @unlocomqx 

I've just tested your PR before it goes to the QA. I noticed a problem with refunds.

In `src/Adapter/Order/Refund/OrderRefundCalculator.php`, there's a method `flattenCheckedProductRefunds`. 

It fails here:

```
if ($quantity > $quantityLeft) {
                throw new InvalidCancelProductException(InvalidCancelProductException::QUANTITY_TOO_HIGH, $quantityLeft);
            }
```

It checks for the quantity in the old way if I'm not mistaken:

```
$customization = new Customization($orderDetail->id_customization);
                $quantityLeft = (int) $customization->quantity - (int) $customization->quantity_refunded - $customization->quantity_returned;
```

This got me thinking, shouldn't we deprecate `Customization::$quantity` as well?

unlocomqx committed:
Deprecate Customization::$quantity

unlocomqx commented:
> shouldn't we deprecate `Customization::$quantity` as well?

Yes, I made the change. 

Customization qty was ignored since commit https://github.com/PrestaShop/PrestaShop/pull/12422/commits/18852a7a2acdce22cc82c816a51106991fdb8a59

kpodemski commented:
@unlocomqx unused `use Customization;` in order refund calculator, other than that we're good to go 🚀

unlocomqx commented:
I fixed the imports in https://github.com/PrestaShop/PrestaShop/commit/8464e9259091cfe0e6a387c724b0b9af4801dac3

Thank you

MatShir reviewed (dismissed):
¡Vamos!

MatShir commented:
Thanks for the work @unlocomqx 🥇

kpodemski commented:
Auto tests:
https://github.com/kpodemski/ga.tests.ui.pr/actions/runs/4086076931

kpodemski commented:
@PrestaShop/qa-team 

Failing auto tests are not related to the PR. Let us know if you want to test something on top of extra tests.

hibatallahAouadni reviewed (commented):
Hello @unlocomqx 

The issue #10912 is already fixed on develop as you can see in the attached screenshot below:

![image](https://user-images.githubusercontent.com/23308427/217559247-69ebab34-be4a-4748-bcd9-be9cba6f2095.png)

While with your two PRs, the customized product image and the Total price are not displayed, and there's some notices displayed above the table, as you can see in the attached screenshot below:

![image](https://user-images.githubusercontent.com/23308427/217559916-7930ca16-8172-4391-bb89-2242127167fb.png)

Please check and feedback.

Thanks!

kpodemski commented:
Hi @hibatallahAouadni 

I've just tested it, and I don't have any errors you reported. I ordered a product with text + image customization, and the same product was added to the cart two times with different customization values.

unlocomqx commented:
@kpodemski I made some tests [here](https://github.com/PrestaShop/classic-theme/pull/25#issuecomment-1422840761)

The issue only occurs when you're on the `develop` branch

kpodemski commented:
Hi @unlocomqx 

I tested your PR thoroughly, and these issues with notices are legit. @hibatallahAouadni tested it correctly :) (as always :D)

`OrderLazyArray`:

```
$productPrice = $includeTaxes ? 'product_price_wt' : 'product_price';
            if (is_array($orderProduct['customizedDatas']) && count($orderProduct['customizedDatas'])) {
                $totalPrice = $includeTaxes ? 'total_customization_wt' : 'total_customization';
            } else {
                $totalPrice = $includeTaxes ? 'total_wt' : 'total_price';
            }
```

You changed this in other places. We should rely on `total_wt` and `total_price` there, right?

This is the last remaining thing to complete. The problem with images is not related to this PR. Customization images are well displayed. I think there is a condition that we don't display images if the product is customizable. Otherwise, we'd show the wrong image if we ordered customizable mugs. I think that was how somebody thought about the current behavior.

unlocomqx commented:
Yes, my bad, I didn't rebase my branch locally

I will revert that line (it was modified to fix a price display bug, but my PR already solves that issue)

unlocomqx committed:
Revert price display logic in confirmation page

We no longer need total_customization_wt

kpodemski reviewed (approved):
Awesome. Let's wait for the tests to be green and we're ready for re-check from QA :)

hibatallahAouadni reviewed (approved):
Hello @unlocomqx 

Thanks for your patience and hard working :pray: 
Finally it works like a charm :raised_hands: :clap: 
LGTM, QA :heavy_check_mark: 

[pr_12422.webm](https://user-images.githubusercontent.com/23308427/217854613-bc3918db-7468-4a3d-990c-1e1e30a888d8.webm)

PS: automated tests are launched : https://github.com/hibatallahAouadni/testing_pr/actions/runs/4135636932

Thanks!

Hlavtox commented:
Thanks everybody!!! Amazing work @unlocomqx! 🚀

unlocomqx commented:
Thanks everyone! always great to contribute 💯 💯
